### ISSUE TITLE
[clang] the error preprocessing directive doesn't handle whitespace and multiline comments

### ISSUE THREAD

**Issue by yakubin** [llvm/llvm-project#63815]
According to the C11 spec, whitespace is handled in translation phase 3, i.e. before the preprocessor runs. Consequently, one can write preprocessing directives spanning multiple lines, as long as the line feed characters are enclosed in multiline comments (`/* ... */`). E.g. this is valid C11:

```
#if  /*
		 */ 1
int main() {
	return 0;
}

#else

ffsdfsdfsdf

#endif
```

And clang handles it correctly.

However, `#error` preprocessor directives, unlike other directives, in Clang seem to not skip whitespace at all. E.g. this directive in Clang preserves all the duplicated spaces in the printed diagnostic:

```
#error one         two
```

(GCC doesn't preserve duplicated spaces in its output.)

And the following program results in two compilation errors instead of one:
```
#error one two /*
	*/ three
```
Clang, when compiling this program, prints:
```
<source>:1:2: error: one two /*
#error one two /*
 ^
<source>:2:3: error: expected identifier or '('
        */ three
         ^
2 errors generated.
```

While GCC prints:
```
<source>:1:2: error: #error one two three
    1 | #error one two /*
      |  ^~~~~
Compiler returned: 1
```

I believe GCC is correct here, since, according to the spec, whitespace (including comments) should be handled before the preprocessor runs, and according to the spec (6.10.5), the `#error` directive accepts tokens (not raw input until end of line).

I compiled all code using ` clang -std=c11`. Locally I used Clang 13 (shipped with XCode). Tested on godbolt.org that the issue is present in Clang 16 as well.

**Comment by shafik** [comment]
CC @AaronBallman

**Comment by AaronBallman** [comment]
Agreed, though this is an intentional behavior in Clang: https://github.com/llvm/llvm-project/blob/2b2e7f6e5727d523ca595ee85a6b1592c8609b4a/clang/lib/Lex/PPDirectives.cpp#L1642

I believe we did that specifically to silence warnings about forming invalid pp-tokens, as in https://godbolt.org/z/r8GGrG5eP, because you run into invalid preprocessor tokens very, very easily otherwise. e.g.,
```
#error please don't do whatever this was doing
```
will give an error but also a warning about `don't` because of the lack of a closing single quote.

I'm confirming the issue because Clang's behavior doesn't match the standard, but this might be a case where we purposefully don't want to match the standard.

Perhaps supporting evidence for Clang's current behavior: https://sourcegraph.com/search?q=context:global+%23error%5Cs%2B%5B%5E%5C%22%5D.*%5B%5C%27%5D%2B.*+lang:C+lang:C%2B%2B&patternType=standard&sm=1&groupBy=repo
